src/app/database/services/database.service/LRUCache.ts

```diff
@@ -41,21 +41,61 @@ export class LRUCache<T> {
         return;
       }
 
-      while (this.currentSize + size > this.size) {
-        const evictedKey = this.lruList.shift();
+      await this.evictEntriesIfNeeded(size);
 
-        if (evictedKey !== undefined) {
-          this.currentSize -= await this.cache.getSize(evictedKey);
-          this.cache.delete(evictedKey);
-        }
-      }
       this.lruList.push(key);
       this.currentSize += size;
       this.cache.set(key, value, size);
-      this.cache.updateLRUState({
-        lruKeyList: this.lruList.slice(),
-        itemsListSize: this.currentSize,
-      });
+      this.persistState();
+    }
+  }
+
+  private async evictEntriesIfNeeded(requiredSize: number): Promise<void> {
+    let hasReconciled = false;
+    let hasEvicted = false;
+
+    while (this.currentSize + requiredSize > this.size) {
+      if (!this.lruList.length) {
+        if (hasReconciled) {
+          break;
+        }
+        await this.reconcileState();
+        hasReconciled = true;
+        if (!this.lruList.length) {
+          break;
+        }
+      }
+
+      const evictedKey = this.lruList.shift();
+
+      if (evictedKey === undefined) {
+        if (!hasReconciled) {
+          await this.reconcileState();
+          hasReconciled = true;
+        }
+        continue;
+      }
+
+      const isCached = await this.cache.has(evictedKey);
+
+      if (!isCached) {
+        if (hasReconciled) {
+          continue;
+        }
+        await this.reconcileState();
+        hasReconciled = true;
```

Comment:
One more suggestion to improve maintainability: consider extracting the eviction logic from `evictEntriesIfNeeded` to reduce its cyclomatic complexity (currently 10).
Could be a method like `tryEvictNext` that:
  - Handles a single eviction attempt (this.lruList.shift())
  - Encapsulates the edge cases (empty list, undefined shift, non-existent cache entry)

This changes I think that will improve: 
  - Single Responsibility Principle: each method has one clear purpose
  - Lower cyclomatic complexity per method 
  - Easy to mantain :)